src/power.jl

```diff
@@ -71,29 +61,50 @@ function simulate_waldtests(
     nβ, m = length(β), deepcopy(morig)
     # we need to do for in-place operations to work across threads
     m_threads = [m]
+    coef_threads = [similar(β)]
+    stderr_threads = [similar(β)]
+    zstat_threads = [similar(β)]
+    pval_threads = [similar(β)]
+    iter_threads = [similar(β, Int)]
 
     if use_threads
         Threads.resize_nthreads!(m_threads)
+        Threads.resize_nthreads!(coef_threads)
+        Threads.resize_nthreads!(stderr_threads)
+        Threads.resize_nthreads!(zstat_threads)
+        Threads.resize_nthreads!(pval_threads)
+        Threads.resize_nthreads!(iter_threads)        
     end
 
+    χsq = Chisq(1)
+
+    cnms = Symbol.(coefnames(morig))
+
+    tidynms = (:coefname,:iteration,:β,:se,:z,:p)
+    results = sizehint!(NamedTuple{tidynms,Tuple{Symbol,Int,T,T,T,T}}[], n)
+
     rnglock = ReentrantLock()
+    pushlock = ReentrantLock()
+    iter = 0
     replicate(n, use_threads=use_threads) do
         mod = m_threads[Threads.threadid()]
         lock(rnglock)
+        iter += 1
+        fill!(iter_threads[Threads.threadid()], iter)
         mod = simulate!(rng, mod, β = β, σ = σ, θ = θ)
         unlock(rnglock)
         refit!(mod)
-        ct = coeftable(mod)
-        names = Tuple(Symbol.(ct.rownms))
-        (
-        # ct.testvalcol
-        # ct.pvalcol
-         β = NamedTuple{names}(ct.cols[1]),
-         se = NamedTuple{names}(ct.cols[2]),
-         z = NamedTuple{names}(ct.cols[3]),
-         p = NamedTuple{names}(ct.cols[4]),
-        )
+        simβ = MixedModels.fixef!(coef_threads[Threads.threadid()], mod)
+        simse = MixedModels.stderror!(stderr_threads[Threads.threadid()], mod)
+        simz = map!(/, zstat_threads[Threads.threadid()], simβ, simse)
+        simp = map!(z -> ccdf(χsq, abs2(z)), pval_threads[Threads.threadid()], simz)
+        lock(pushlock)
+        for t in zip(cnms, iter_threads[Threads.threadid()], simβ, simse, simz, simp)
+            push!(results, NamedTuple{tidynms}(t))
+        end
+        unlock(pushlock)
```

Comment:
Hmmm, since we have the index with `iter`, it may be possible to avoid locking here. If we we know that each iteration yields `p` betas, then we can pre allocate for `p * n_iter1` and just do normal assignment to `p * (iter - 1) + idx` for `1 <= idx <= p`. In addition to being a point of congestion, the (un)locking process itself isn't free.